### PR TITLE
11 Files 

### PR DESCRIPTION
**Pull request by EmmanuelStalling**

### CODE REVIEW

File: firebase-deploy-react-js.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+After we built a full-fledged Firebase application in React, the final step is deployment, the tipping point of getting your ideas out into the world, from consuming tutorials to producing applications. Since you have already used Firebase extensively for your application, why not choosing Firebase Hosting for the deployment?
+
+In this section, I want to guide you through deploying your React to Firebase. It works for create-react-app too. Also it should work for any other library and framework such as Angular or Vue. First, install the Firebase CLI globally to your node modules:
```
Comment (rwieruch) [inline firebase-deploy-react-js.md]: "deploying your React to" -> "deploying your React application to"

File: firebase-deploy-react-js.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+{{< highlight javascript >}}
+npm install -g firebase-tools
+{{< /highlight >}}
+
+Using a global installation of the Firebase CLI, you can deploy any application without worrying the dependency in your project. For any global installed node package, remember to update the it occasionally to a newer version with the identical command:
```
Comment (rwieruch) [inline firebase-deploy-react-js.md]: "worrying the dependency" -> "worrying about the dependency"

Comment (rwieruch) [reply firebase-deploy-react-js.md]: "update the it" -> "update it"

File: firebase-deploy-react-js.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+There are a few other configuration steps to define. Instead of using the default *public/* folder, we want to use the *build/* folder for create-react-app. If you set up the bundling with a tool like Webpack, you can choose the appropriate name for the build folder:
+
+{{< highlight javascript >}}
+? What do you want to use as your public directory? build
+? Configure as a single-page app (rewrite all urls to /index.html)? Yes
+? File public/index.html already exists. Overwrite? No
+{{< /highlight >}}
+
+The create-react-app application creates a *build/* folder after you perfrom the`npm run build` for the first time. There you will find all the merged content from the *public/* folder and the *src/* folder. Since it is a single page application, we want to redirect the user always to the *index.html* file. From there React Router takes over for the client-side routing.
```
Comment (rwieruch) [inline firebase-deploy-react-js.md]: "perfrom the`npm" <- missing space

File: react-firebase-auth-persistence.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    ...
+  }
+
+  ...
+};
+
+...
+{{< /highlight >}}
+
+Every time Firebase's listener is invoked, the authenticated user is not only stored in the local state, ready to be passed to React's Context API, but  it's also stored in the browser's local storage. You can use the local storage's API with `setItem` and `removeItem` to store and delete something identified by a key. You also need to format the authenticated user to JSON before you can put it into the local storage of the browser.
```
Comment (rwieruch) [inline react-firebase-auth-persistence.md]: "but  it" <- two spaces

File: react-firebase-email-verification.md [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+
+In your application, users can employ an email/password combination, but also social logins to get access to your service or product. Often, the email address associated with the social logins is confirmed by the social platform (Google, Facebook, Twitter) and you know this email address really exists. But what about the email address used with the password? Because users are sometimes unwilling to provide real email addresses, they'll simply make one  up, so you can't provide them with further information via email or to integrate them with third-parties where a valid email address is required. In this section, I will show you how to confirm user email addresses before they can access your application. After an email verification with a double opt-in send by email, users are authorized to use your application.
```
Comment (rwieruch) [inline react-firebase-email-verification.md]: "one  up" <- double spaces

File: react-firebase-email-verification.md [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+      }
+    });
+
+    ...
+}
+
+export default Firebase;
+{{< /highlight >}}
+
+To protect your routes from users who have no verified email address? We will do it with a new higher-order component in *src/components/Session/withEmailVerification.js* that has access to Firebase and the authenticated user:
```
Comment (rwieruch) [inline react-firebase-email-verification.md]: "To protect your routes from users who have no verified email address?" <- should this be a question?

File: react-firebase-email-verification.md [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+{{< highlight javascript "hl_lines=1 2 3 4 5 6" >}}
+const needsEmailVerification = authUser =>
+  authUser &&
+  !authUser.emailVerified &&
+  authUser.providerData
+    .map(provider => provider.providerId)
+    .includes('password');
+{{< /highlight >}}
+
+If this is true, don't render the component passed to this higher-order component, but a message that reminds users to verify t their email addresses.
```
Comment (rwieruch) [inline react-firebase-email-verification.md]: "verify t their" -> "verify their"

File: react-firebase-email-verification.md [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+    }
+  }
+
+  return withFirebase(WithEmailVerification);
+};
+
+export default withEmailVerification;
+{{< /highlight >}}
+
+Optionally, we can provide a button to resend a verification email to the user. Let's improve the user experience. After the button is clicked to resend the verification email, users should receive feedback, and be prohibited from sending b another email. First, add a local state to the higher-order component that tracks whether the button was clicked:
```
Comment (rwieruch) [inline react-firebase-email-verification.md]: "sending b another" -> "sending another"

File: react-firebase-email-verification.md [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+    }
+  }
+
+  return withFirebase(WithEmailVerification);
+};
+
+export default withEmailVerification;
+{{< /highlight >}}
+
+
```
Comment (rwieruch) [inline react-firebase-email-verification.md]: two empty lines -> one empty line

File: react-firebase-email-verification.md [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+const condition = authUser => !!authUser;
+
+export default compose(
+  withEmailVerification,
+  withAuthorization(condition),
+)(AccountPage);
+{{< /highlight >}}
+
+All the sensible routes for authenticated users now require a confirmed email. Finally,m your application can be only used by users with read email addresses.
```
Comment (rwieruch) [inline react-firebase-email-verification.md]: Finally,m -? Finally,

Comment (rwieruch) [reply react-firebase-email-verification.md]: read -> real

File: react-firebase-link-social-logins.md [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+    event.preventDefault();
+  };
+
+  ...
+}
+
+...
+{{< /highlight >}}
+
+Now users can use the same email address for different sign-in methods.  Next, et's head to the account page, where we'll create an area to manage and activate/deactivate all the sign-in methods (social sign-ins, email/password sign-in). Introduce all available sign-in methods and their optional providers (see Firebase class) as list of objects:
```
Comment (rwieruch) [inline react-firebase-link-social-logins.md]: "methods.  Next, et's" -> "methods. Next, let's"

File: react-firebase-link-social-logins.md [hunk 12]
```diff
@@ -0,0 +1,619 @@
...
+          })}
+        </ul>
+        {error && error.message}
+      </div>
+    );
+  }
+}
+{{< /highlight >}}
+
+Also, we added an improvement to avoid getting locked out of the application. If only one sign-in method is left as activate, disable all deactivation buttons because there needs to be at least one sign-in method. Now let's implement the class methods for linking and unlinking accounts:
```
Comment (rwieruch) [inline react-firebase-link-social-logins.md]: "left as activate" -> "left as active" ?

File: react-firebase-mobx-tutorial.md [hunk 13]
```diff
@@ -0,0 +1,842 @@
...
+        {user && ( ... )}
+      </div>
+    );
+  }
+}
+{{< /highlight >}}
+
+That's it for the UserItem component. It renders a user, fetches the recent user from Firebase with a realtime connection, only it stores the result into the MobX store.
+
+<hr class="section-divider">
```
Comment (rwieruch) [inline react-firebase-mobx-tutorial.md]: Remove this line.

File: react-firebase-realtime-database.md [hunk 14]
```diff
@@ -0,0 +1,1110 @@
...
+);
+
+const MessageItem = ({ message }) => (
+  <li>
+    <strong>{message.userId}</strong> {message.text}
+  </li>
+);
+{{< /highlight >}}
+
+If you run the application, the loading indicator disappears after a few seconds when the Firebase realtime database listener is called for the first time. Every other time the loading indicator isn't shown, because it is only`true` when the component mounts and the first message fetching starts.
```
Comment (rwieruch) [inline react-firebase-realtime-database.md]: "only`true" -> "only `true"

File: react-firebase-realtime-database.md [hunk 15]
```diff
@@ -0,0 +1,1110 @@
...
+
+  ...
+}
+{{< /highlight >}}
+
+Since you have no messages, nothing shows up. Creating chat messages is our next task.
+
+{{% chapter_header "Create an Item in a List in Firebase Realtime Database" "firebase-realtime-database-create-item-in-list" %}}
+
+We were able to get all messages from the Firebase realtime database. It's even pdated for us using the Firebase listener on a reference with the `on` and not `once` method. Next, let's implement a React form that lets us create a message entity in the Firebase realtime database:
```
Comment (rwieruch) [inline react-firebase-realtime-database.md]: pdated -> updated

File: react-firebase-redux-tutorial.md [hunk 16]
```diff
@@ -0,0 +1,926 @@
...
+    });
+  }
+
+  ...
+}
+{{< /highlight >}}
+
+The users are no longer managed in the local state of the component, but are now handled in Redux. You set the users with a dispatchable action from `mapDispatchToProps` and access them again in `mapStateToProps`. Both state and actions are passed as props to your component.
+
+The users and loading indicator are rendered as before, but only the loading statecomes from the local state.The Link component only navigates to the UserItem component, but it doesn't send any user objects. We wanted the user at our disposal via the Link component, and we want to let Redux handle it.
```
Comment (rwieruch) [inline react-firebase-redux-tutorial.md]: statecomes -> state comes

Comment (rwieruch) [reply react-firebase-redux-tutorial.md]: state.The -> state. The

File: react-firebase-redux-tutorial.md [hunk 17]
```diff
@@ -0,0 +1,926 @@
...
+export default compose(
+  withFirebase,
+  connect(
+    mapStateToProps,
+    mapDispatchToProps,
+  ),
+)(UserItem);
+{{< /highlight >}}
+
+Similar to the UserList component,treceives a function that is a dispatchable action that sets a user instead of users. Check the user reducer to see what's happening when this action is dispatched. The component receives a single user from the Redux store. Because the initial state of users in the Redux store is null, we have to conditionally choose an empty object for not running into null pointer exceptions.
```
Comment (rwieruch) [inline react-firebase-redux-tutorial.md]: treceives?

File: react-firebase-redux-tutorial.md [hunk 18]
```diff
@@ -0,0 +1,926 @@
...
+    );
+  }
+}
+{{< /highlight >}}
+
+That's it for the UserItem component. It renders a user, fetches the recent user from Firebase with a realtime connection, but stores the result into the Redux store.
+
+<hr class="section-divider">
+
+The advantage of using Redux instead of React's local state a persistent state of users between routes. That means you don't need to fetch the users every time you navigate from UserItem to UserList or any other route, because they remain in Redux's global state after they are fetched the first time.
```
Comment (rwieruch) [inline react-firebase-redux-tutorial.md]: "state a persistent" -> "state is a persistent"

File: react-firestore-tutorial.md [hunk 19]
```diff
@@ -0,0 +1,581 @@
...
+
+* {{% a_blank "React with Firebase Realtime Database" "https://github.com/the-road-to-react-with-firebase/react-firebase-authentication" %}}.
+* {{% a_blank "React with Firebase Cloud Firestore" "https://github.com/the-road-to-react-with-firebase/react-firestore-authentication" %}}.
+
+Except for the database, everything else stays the same; thus, everything else you learned from my previous React Firebase tutorials is still up-to-date. Before we start with migration, consider reading through this [visual Firebase tutorial](https://www.robinwieruch.de/firebase-tutorial/) to set up your Firebase project with Cloud Firestore.
+
+{{% chapter_header "Migration from Realtime Database to Cloud Firestore" "firebase-cloud-firestore-migration" %}}
+
+First, our project has a Firebase class that connects our React application with the Firebase API (e.g. authentication API, database API). It currently 0uses Firebase's Realtime Database:
```
Comment (rwieruch) [inline react-firestore-tutorial.md]: 0uses -> uses

File: react-firestore-tutorial.md [hunk 20]
```diff
@@ -0,0 +1,581 @@
...
+
+  render() {
+    ...
+  }
+}
+{{< /highlight >}}
+
+However, as seen in the code snippet, Cloud Firestore comes with a merge option. If you are not sure whether your document already exists, pass the merge option to avoid overwriting the entire document. New content is merged into the entity if the entity is already there. We don't use the `update()` method because it fails if the document doesn't exist.
+
+In our case, the merge operation makes sense because we can't be sure if a user is signing up for the first time  or if they've signed up with a social login such as Google or Facebook. To handle this, we migrate our user creations for the social logins in the sign in form to use the merge option too.
```
Comment (rwieruch) [inline react-firestore-tutorial.md]: time  or <- double spaces

File: react-firestore-tutorial.md [hunk 21]
```diff
@@ -0,0 +1,581 @@
...
+        <p>The Home Page is accessible by every signed in user.</p>
+
+        <Messages users={this.state.users} />
+      </div>
+    );
+  }
+}
+{{< /highlight >}}
+
+In this component, we don't want to transform the collection into a list of items but keep it as object that can be accessed using identifiers in the form of a dictionary later. The Messages component  fetches our other collection that we adjusted in the beginning in the Firebase class:
```
Comment (rwieruch) [inline react-firestore-tutorial.md]: component  fetches <- double spaces

File: react-firebase-email-verification.md (hunk 5 above)
Comment (EmmanuelStalling) [reply react-firebase-email-verification.md]: I don't see the phrase "one up" in the text, not could I find it with a command+F search.

File: react-firebase-redux-tutorial.md (hunk 17 above)
Comment (EmmanuelStalling) [reply react-firebase-redux-tutorial.md]: it receives

File: react-firebase-realtime-database.md [hunk 22]
```diff
@@ -0,0 +1,1110 @@
...
+      .on('value', snapshot => {
+        ...
+      });
+  }
+
+  ...
+}
+{{< /highlight >}}
+
+Limiting the items is half the task for enabling pagination for our chat application. We also need to move the limit to the local state of the component to adjust it later with user interactions to fetch more than five(5) items:
```
[the remark this answers is not in the document]
Comment (EmmanuelStalling) [reply react-firebase-realtime-database.md]: Always write out the number in technical writing, unless the number represents a quantity and a value to use with a variable, at which point it's both.

File: react-firebase-email-verification.md (hunk 5 above)
Comment (rwieruch) [reply react-firebase-email-verification.md]: "they'll simply make one  up, so" <- I can fix it myself 👍

**Pull request merged**